Font fallback

jonleighton opened the pull request:
This implements more complete support for font fallback, see #17267.

r? @glennw @mbrubeck

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/20506)
<!-- Reviewable:end -->

jonleighton committed:
Implement font fallback

Prior to this change, if none of the fonts specified in CSS contained a
glyph for a codepoint, we tried only one fallback font. If that font
didn't contain the glyph, we'd give up.

With this change, we try multiple fonts in turn. The font names we try
differ across each platform, and based on the codepoint we're trying to
match. The current implementation is heavily inspired by the analogous
code in Gecko, but I've used to ucd lib to make it more readable,
whereas Gecko matches raw unicode ranges.

This fixes some of the issues reported in #17267, although colour emoji
support is not implemented.

== Notes on changes to WPT metadata ==

=== css/css-text/i18n/css3-text-line-break-opclns-* ===

A bunch of these have started failing on macos when they previously
passed.

These tests check that the browser automatically inserts line breaks
near certain characters that are classified as "opening and closing
punctuation". The idea is that if we have e.g. an opening parenthesis,
it does not make sense for it to appear at the end of a line box; it
should "stick" to the next character and go into the next line box.

Before this change, a lot of these codepoints rendered as a missing
glyph on Mac and Linux. In some cases, that meant that the test was
passing.

After this change, a bunch of these codepoints are now rendering glyphs
on Mac (but not Linux). In some cases, the test should continue to pass
where it previously did when rendering with the missing glyph.

However, it seems this has also exposed a layout bug. The "ref" div in
these tests contains a <br> element, and it seems that this, combined
with these punctuation characters, makes the spacing between glyphs ever
so slightly different to the "test" div. (Speculation: might be
something to do with shaping?)

Therefore I've had to mark a bunch of these tests failing on mac.

=== css/css-text/i18n/css3-text-line-break-baspglwj-* ===

Some of these previously passed on Mac due to a missing glyph. Now that
we're rendering the correct glyph, they are failing.

=== css/css-text/word-break/word-break-normal-bo-000.html ===

The characters now render correctly on Mac, and the test is passing. But
we do not find a suitable fallback font on Linux, so it is still failing
on that platform.

=== css/css-text/word-break/word-break-break-all-007.html ===

This was previously passing on Mac, but only because missing character
glyphs were rendered. Now that a fallback font is able to be found, it
(correctly) fails.

=== mozilla/tests/css/font_fallback_* ===

These are new tests added in this commit. 01 and 02 are marked failing
on Linux because the builders don't have the appropriate fonts installed
(that will be a follow-up).

Fix build errors from rebase

FontTemplateDescriptor can no longer just derive(Hash). We need to
implement it on each component part, because the components now
generally wrap floats, which do not impl Hash because of NaN. However in
this case we know that we won't have a NaN, so it is safe to manually
impl Hash.

jonleighton committed:
FreeType: Improve "Invalid codepoint" debug message

jonleighton committed:
Linux: Don't hold onto bytes of system fonts

FontTemplateData gets passed over IPC during the communication between
FontContext and FontCacheThread. Serializing and deserializing these
bytes is expensive, so this change ensures that we only do that when the
bytes can't be read from disk. A similar strategy is already used on
macos and windows.

The performance problem was particularly noticeable after implenting
font fallback, where the content process would potentially work through
a list of fonts, trying to find one which contains a certain glyph. That
could result in lots of font bytes going over IPC.

jonleighton committed:
FontContext: Cache data fetched from the cache thread

Before this change, if we needed to create a Font which we've already
created, but at a new size, then we'd fetch the FontTemplateInfo again.
If the bytes of the font are held in memory, then this could be
expensive as we need to pass those bytes over IPC.

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/layout/text.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

jdm commented:
This looks amazing! \o/

glennw commented:
Reviewed 17 of 23 files at r1.
Review status: 17 of 23 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[components/gfx/platform/freetype/font.rs, line 102 at r1](https://beta.reviewable.io/reviews/servo/servo/20506#-L9CkswM0vMeJ26HgbGv:-L9CkswM0vMeJ26HgbGw:b-ihiv8a) ([raw file](https://github.com/servo/servo/blob/e15d3e11c35503df0390811e2c14e393d6e90304/components/gfx/platform/freetype/font.rs#L102)):*
> ```Rust
>             FT_New_Memory_Face(lib, bytes.as_ptr(), bytes.len() as FT_Long, face_index, &mut face)
>         } else {
>             let filename = CString::new(&*template.identifier).expect("filename contains NUL byte!");
> ```

One of the reasons for making FT always load via memory here was so that the layout threads would never block on a synchronous file load. The idea was that the font cache thread would responsible for providing these. I'm not sure if that's really relevant or feasible, but that was the original idea, I think.

---

*[components/gfx/text/util.rs, line 156 at r1](https://beta.reviewable.io/reviews/servo/servo/20506#-L9CikUY45TvQYJOiwZN:-L9CikUY45TvQYJOiwZO:b-gmualv) ([raw file](https://github.com/servo/servo/blob/e15d3e11c35503df0390811e2c14e393d6e90304/components/gfx/text/util.rs#L156)):*
> ```Rust
> 
>     // https://en.wikipedia.org/wiki/Plane_(Unicode)#Supplementary_Ideographic_Plane
>     if unicode_plane(codepoint) == 2 {
> ```

nit: could just be unicode_plane(codepoint) == 2 and remove the true / false here, unless there's more conditions to come in the future.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/servo/servo/20506)*
<!-- Sent from Reviewable.io -->

glennw commented:
Partial review only so far - I'll try to work through the trickier bits today or tomorrow. Will need @mbrubeck to take a look at some of the text files that I'm not familiar with too.

glennw commented:
Reviewed 6 of 23 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/layout/text.rs, line 218 at r1](https://beta.reviewable.io/reviews/servo/servo/20506#-L9HkoS_Ax-B3TmtGbNc:-L9HkoS_Ax-B3TmtGbNd:b-nd2use) ([raw file](https://github.com/servo/servo/blob/e15d3e11c35503df0390811e2c14e393d6e90304/components/layout/text.rs#L218)):*
> ```Rust
>                 let (mut start_position, mut end_position) = (0, 0);
>                 for (byte_index, character) in text.char_indices() {
>                     if !character.is_control() {
> ```

Should get @mbrubeck to check this.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/servo/servo/20506)*
<!-- Sent from Reviewable.io -->

glennw commented:
Reviewed the remaining files, they look good to me. I think it'd be worth getting @mbrubeck to do a quick pass over this too, as I'm not very up to date on text / font handling in Servo. My main question is related to the thread the font load occurs on (mentioned above).

jonleighton commented:
> One of the reasons for making FT always load via memory here was so that the layout threads would never block on a synchronous file load. The idea was that the font cache thread would responsible for providing these. I'm not sure if that's really relevant or feasible, but that was the original idea, I think.

Thanks for the explanation. I note that the windows and macos implementations end up calling into platform APIs in this situation, do you know of calling those APIs results in any disk I/O?

Without this change I observed a big slowdown on Linux where the bytes of all the fallback fonts would be passed over the ipc-channel. The slowdown was mostly due to the CPU cost of serialization and deserialization. Performance was a lot better when I tried [the `serde_bytes` crate](https://crates.io/crates/serde_bytes), so that could be an option here (actually I think it is worth using `serde_bytes` regardless, but it may make performance acceptable if we want to keep all the I/O in the font cache thread).

One thing it is worth pointing out, is that by the time a layout thread does the file read, the file has already been read by the font cache thread. Therefore, we might hope that the data is already in an OS cache and it might be that no actual disk reads need to occur. I don't know enough about how this works to make any firm statements about that though. Another thing to say is that I'm testing this on a laptop with an SSD hard drive, so I'm probably less likely to notice disk I/O slowdowns.

Do you have any suggestions about how I could/should investigate this further so we can decide what the best approach is?

jonleighton commented:
I should add: when I looked into the bytes-over-IPC thing, I realised that the font code had originally been developed before a multiprocess architecture was implemented. Therefore, at the time, those bytes were not getting serialized or sent to a different process. Then, the multiprocess architecture came along and converted everything to IPC channels. I thought that perhaps it wasn't intentionally decided that the bytes should go over IPC, but instead just a consequence of how the code had evolved. It just seemed quite surprising to me to do this, but it might be that it's actually fine if we use `serde_bytes`.

glennw commented:
You're exactly right - the original font code was built before any consideration for multi-process / ipc-channel, and serializing the bytes over the ipc channel was a side effect of that.

I think it's fine to move the file read into the font contexts at this stage - this is certainly a big improvement on what we have now.

We can re-consider the design of how / where we fetch the bytes as a separate issue for later - it's a big topic by itself.

There's no other blockers from me, but it would be good to get @mbrubeck to do a quick pass over this before merging - he has much more experience with the Servo text / layout code than me.

mbrubeck commented:
@bors-servo r+

Looks good to me.  Thanks for this!

bors-servo commented:
:pushpin: Commit e15d3e1 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck e15d3e11c35503df0390811e2c14e393d6e90304 -->

bors-servo commented:
:hourglass: Testing commit e15d3e11c35503df0390811e2c14e393d6e90304 with merge b278ee4c164a15601dc7ab7e007384a99a9611b5...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11590)

mbrubeck commented:
```
error: unused import: `std::borrow::ToOwned`
 --> components/gfx/platform/macos/font_list.rs:6:5
  |
6 | use std::borrow::ToOwned;
  |     ^^^^^^^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
```

jonleighton committed:
Don't perform font matching for control characters

We can encounter control characters here, for example when processing a
<pre> element which contains newlines. Control characters are inherently
non-printing, therefore if we try to call find_by_codepoint for these
characters we will end up triggering an unnecessary font fallback
search.

jonleighton commented:
> nit: could just be unicode_plane(codepoint) == 2 and remove the true / false here, unless there's more conditions to come in the future.

Done

> Should get @mbrubeck to check this.

This wasn't actually 100% necessary for the fallback to work, though it's useful in terms of not doing unnecessary fallbacks for non-printing characters. I've split it out into a separate commit which gives a better explanation of what it does.

> We can re-consider the design of how / where we fetch the bytes as a separate issue for later - it's a big topic by itself.

I've added a comment to the code linking to this discussion.

> error: unused import: `std::borrow::ToOwned`

Fixed

mbrubeck commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 6679ebe has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 6679ebe24b72d6cee87a536739903e26f90ebc20 -->

bors-servo commented:
:hourglass: Testing commit 6679ebe24b72d6cee87a536739903e26f90ebc20 with merge 8a75598f379558c9121f263149c515a927d3ee42...

jonleighton commented:
Fixed that one, sorry

mbrubeck commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 728b57b has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 728b57b5acb054e697dd4fc0aac55af6230a26d0 -->

bors-servo commented:
:hourglass: Testing commit 728b57b5acb054e697dd4fc0aac55af6230a26d0 with merge 3c3275849dfd595fe7ae7819218eb2020eccee69...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8427)

CYBAI commented:
```
{"status": "TIMEOUT", "group": "default", "message": "Test timed out", "stack": null, "subtest": "Sniffed font is loaded and applied", "test": "/_mozilla/mozilla/mime_sniffing_font_context.html", "line": 136227, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": "Test timed out", "stack": null, "subtest": "Sniffed font with xml is not loaded", "test": "/_mozilla/mozilla/mime_sniffing_font_context.html", "line": 136228, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": "Test timed out", "stack": null, "subtest": "Sniffed font with xhtml+xml is not loaded", "test": "/_mozilla/mozilla/mime_sniffing_font_context.html", "line": 136229, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/mozilla/mime_sniffing_font_context.html", "line": 136233, "action": "test_result", "expected": "OK"}
```

jonleighton commented:
That one was tricky to track down! Turns out I forgot to clear the new `font_template_cache` in `FontContext::expire_font_caches_if_necessary` :flushed:

mbrubeck commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 0f7b633 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 0f7b6336d6bee534305461361ab3436d0c2d319e -->

bors-servo commented:
:hourglass: Testing commit 0f7b6336d6bee534305461361ab3436d0c2d319e with merge 62e57c4b2dac0effcef3d3b9e9fde79741666425...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8447)

mbrubeck commented:
Lots of line-breaking tests seem to be affected.  I haven't investigated to see which of these failures are new bugs, and which are tests that were "accidentally passing" before.

linux-rel-wpt:

```
{"status": "FAIL", "group": "default", "message": "/_mozilla/css/font_fallback_01.html 933650d419a5c77b9aaf59d983187ab712b75312\n/_mozilla/css/font_fallback_failed_ref.html 933650d419a5c77b9aaf59d983187ab712b75312\nTesting 933650d419a5c77b9aaf59d983187ab712b75312 != 933650d419a5c77b9aaf59d983187ab712b75312", "stack": null, "subtest": null, "test": "/_mozilla/css/font_fallback_01.html", "line": 42512, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/_mozilla/css/font_fallback_02.html 933650d419a5c77b9aaf59d983187ab712b75312\n/_mozilla/css/font_fallback_failed_ref.html 933650d419a5c77b9aaf59d983187ab712b75312\nTesting 933650d419a5c77b9aaf59d983187ab712b75312 != 933650d419a5c77b9aaf59d983187ab712b75312", "stack": null, "subtest": null, "test": "/_mozilla/css/font_fallback_02.html", "line": 49681, "action": "test_result", "expected": "PASS"}
```

(Also a number of [other unexpected results](http://build.servo.org/builders/linux-rel-wpt/builds/8447/steps/shell__4/logs/intermittents.log) that got filtered out as intermittent but might be relevant, since several of them are in text layout tests.)

mac-rel-css2:

```
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/css/css-text/word-break/word-break-normal-bo-000.html", "line": 7001, "action": "test_result", "expected": "FAIL"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/word-break/word-break-break-all-007.html d37ece18dd4e84bd8450396b5b8e690bc2968709\n/css/css-text/word-break/reference/word-break-break-all-ref-007.html 1d17bafb661f4d6938e8ef641f84aec83b938a34\nTesting d37ece18dd4e84bd8450396b5b8e690bc2968709 == 1d17bafb661f4d6938e8ef641f84aec83b938a34", "stack": null, "subtest": null, "test": "/css/css-text/word-break/word-break-break-all-007.html", "line": 7593, "action": "test_result", "expected": "PASS"}
```

mac-rel-wpt1:

<details>

```
{"status": "FAIL", "group": "default", "message": "/css/CSS2/fonts/font-family-invalid-characters-001.xht a9b04d155de929d05baa66bd102a34c03658802b\n/css/CSS2/fonts/font-family-invalid-characters-001-ref.html 48bebf0d46817167cf3310a994a28ad3370930dc\nTesting a9b04d155de929d05baa66bd102a34c03658802b == 48bebf0d46817167cf3310a994a28ad3370930dc", "stack": null, "subtest": null, "test": "/css/CSS2/fonts/font-family-invalid-characters-001.xht", "line": 2546, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-014.html 73226ed83d93ed1b659505bff24ca9a9bf1935d5\n/css/css-text/i18n/reference/css3-text-line-break-opclns-014-ref.html 2628a64b519062fad477448b8122aa175eb8d86f\nTesting 73226ed83d93ed1b659505bff24ca9a9bf1935d5 == 2628a64b519062fad477448b8122aa175eb8d86f", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-014.html", "line": 8971, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-018.html d130d77fbbac76bc1bf1f84de33c8fe063427526\n/css/css-text/i18n/reference/css3-text-line-break-opclns-018-ref.html 45a79304107056cd2ebe18693e2e1d61b69d65ab\nTesting d130d77fbbac76bc1bf1f84de33c8fe063427526 == 45a79304107056cd2ebe18693e2e1d61b69d65ab", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-018.html", "line": 8993, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-025.html 9fcc9eaef2b2d3862b6eb53294be47f7bf38801c\n/css/css-text/i18n/reference/css3-text-line-break-opclns-025-ref.html cb4858a17602169fa2d45b30855f4adc36baff7c\nTesting 9fcc9eaef2b2d3862b6eb53294be47f7bf38801c == cb4858a17602169fa2d45b30855f4adc36baff7c", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-025.html", "line": 9036, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-029.html 7ea95a8c9f1fb7bcf8dc2e4664ff3e7ae69ab494\n/css/css-text/i18n/reference/css3-text-line-break-opclns-029-ref.html 6a7ebec23be853db7185efd13c05ec1e44e0315d\nTesting 7ea95a8c9f1fb7bcf8dc2e4664ff3e7ae69ab494 == 6a7ebec23be853db7185efd13c05ec1e44e0315d", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-029.html", "line": 9063, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-032.html 0efdc36ae34a3b0929605cceab0d2951935e1f27\n/css/css-text/i18n/reference/css3-text-line-break-opclns-032-ref.html e54acd35d64b140d40e3dd6feb200302cee54a18\nTesting 0efdc36ae34a3b0929605cceab0d2951935e1f27 == e54acd35d64b140d40e3dd6feb200302cee54a18", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-032.html", "line": 9085, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-036.html 2fb633d11590db236bdf89854c723dbf1de35330\n/css/css-text/i18n/reference/css3-text-line-break-opclns-036-ref.html 6769538d90cc34b60c7fc53a7b96740208effad1\nTesting 2fb633d11590db236bdf89854c723dbf1de35330 == 6769538d90cc34b60c7fc53a7b96740208effad1", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-036.html", "line": 9107, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-026.html 2490f4995305789a84ce8a649d4e7d6f6409ace8\n/css/css-text/i18n/reference/css3-text-line-break-opclns-026-ref.html aa9432937e45e76f2a11197611418c6a923853b2\nTesting 2490f4995305789a84ce8a649d4e7d6f6409ace8 == aa9432937e45e76f2a11197611418c6a923853b2", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-026.html", "line": 9325, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-031.html d37718962becf9f0a40a2fb8b658cf86bc7b8f84\n/css/css-text/i18n/reference/css3-text-line-break-opclns-031-ref.html 7d9de58f2888686fb05ce6abc1fbb2c214bd9a91\nTesting d37718962becf9f0a40a2fb8b658cf86bc7b8f84 == 7d9de58f2888686fb05ce6abc1fbb2c214bd9a91", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-031.html", "line": 9347, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-119.html ed622560e222a01a0a83539025cdf1b4c22e667a\n/css/css-text/i18n/reference/css3-text-line-break-opclns-119-ref.html 3617564ec71fd5fd3981c5073c3226767c384207\nTesting ed622560e222a01a0a83539025cdf1b4c22e667a == 3617564ec71fd5fd3981c5073c3226767c384207", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-119.html", "line": 9355, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-035.html 5a600ca76d15b350c9dd3f7ffd01a31c709eb9f9\n/css/css-text/i18n/reference/css3-text-line-break-opclns-035-ref.html ea27d1216e7ab72339b23803f7759e723dec4d6c\nTesting 5a600ca76d15b350c9dd3f7ffd01a31c709eb9f9 == ea27d1216e7ab72339b23803f7759e723dec4d6c", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-035.html", "line": 9371, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-120.html 3c65f87456b81bb025a54e77ce1ebdfca1ae9397\n/css/css-text/i18n/reference/css3-text-line-break-opclns-120-ref.html f5e08d906dd3ef9a26aea9b305b98d0ec6ae17c5\nTesting 3c65f87456b81bb025a54e77ce1ebdfca1ae9397 == f5e08d906dd3ef9a26aea9b305b98d0ec6ae17c5", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-120.html", "line": 9379, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-124.html 1c583bd8736c26718a6f1992bd0d1829b8ee14ed\n/css/css-text/i18n/reference/css3-text-line-break-opclns-124-ref.html feb76bbdb3b8dd38df3d18fb5ef36b496a53c401\nTesting 1c583bd8736c26718a6f1992bd0d1829b8ee14ed == feb76bbdb3b8dd38df3d18fb5ef36b496a53c401", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-124.html", "line": 9401, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-128.html 3eb09cdd44a0fedd984c4be47d7671d8eea1cab6\n/css/css-text/i18n/reference/css3-text-line-break-opclns-128-ref.html da7b53f83257a1fc34cd8af1537ed1b315c50ff4\nTesting 3eb09cdd44a0fedd984c4be47d7671d8eea1cab6 == da7b53f83257a1fc34cd8af1537ed1b315c50ff4", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-128.html", "line": 9424, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-133.html 2bea738a097b9f1f6ca6bfa70ff4abec6ba3413b\n/css/css-text/i18n/reference/css3-text-line-break-opclns-133-ref.html ba364bc4edf848dada4f73826bcbcd2462dd0f9b\nTesting 2bea738a097b9f1f6ca6bfa70ff4abec6ba3413b == ba364bc4edf848dada4f73826bcbcd2462dd0f9b", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-133.html", "line": 9449, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-137.html f2f8821fabf869f3ca123ba3165b5f23bd0cb59d\n/css/css-text/i18n/reference/css3-text-line-break-opclns-137-ref.html 80b54561cd51cdc3aa687a731e3f854ac7015c66\nTesting f2f8821fabf869f3ca123ba3165b5f23bd0cb59d == 80b54561cd51cdc3aa687a731e3f854ac7015c66", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-137.html", "line": 9471, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-142.html 72ce3a5589dc32644a90bc69a8ad8201389ab1e4\n/css/css-text/i18n/reference/css3-text-line-break-opclns-142-ref.html 02ea661f271dea1c10dc366e089c4fdcad8ebbdc\nTesting 72ce3a5589dc32644a90bc69a8ad8201389ab1e4 == 02ea661f271dea1c10dc366e089c4fdcad8ebbdc", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-142.html", "line": 9493, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-123.html 220bafad839032a3132718be984dbbed41a2f783\n/css/css-text/i18n/reference/css3-text-line-break-opclns-123-ref.html ca57a254c344a06e38f76831b033ac3966f77a1f\nTesting 220bafad839032a3132718be984dbbed41a2f783 == ca57a254c344a06e38f76831b033ac3966f77a1f", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-123.html", "line": 9621, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-130.html e6d80d918db9e517e6b96559740c89a57f4bd473\n/css/css-text/i18n/reference/css3-text-line-break-opclns-130-ref.html 2fce2d6b1872d0904ba453d8b68cdfabf50b20b1\nTesting e6d80d918db9e517e6b96559740c89a57f4bd473 == 2fce2d6b1872d0904ba453d8b68cdfabf50b20b1", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-130.html", "line": 9659, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-134.html 54f50e7296def1a8d91b71d2aec126553ffc2e1c\n/css/css-text/i18n/reference/css3-text-line-break-opclns-134-ref.html 021dadcfd318f83dc597effefbfdd1c3e8c96512\nTesting 54f50e7296def1a8d91b71d2aec126553ffc2e1c == 021dadcfd318f83dc597effefbfdd1c3e8c96512", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-134.html", "line": 9682, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-138.html 890b47e8741eed2ec8d5a3ade02608ab8799b395\n/css/css-text/i18n/reference/css3-text-line-break-opclns-138-ref.html 6b1e3910ed20a13e1aba6aba7496d99c4654e8d3\nTesting 890b47e8741eed2ec8d5a3ade02608ab8799b395 == 6b1e3910ed20a13e1aba6aba7496d99c4654e8d3", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-138.html", "line": 9704, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-141.html f544ec76c810ba735b5d1492c493e84cfe81a719\n/css/css-text/i18n/reference/css3-text-line-break-opclns-141-ref.html 43dc4f85a9519e9c93e65cba75271725a8f17e51\nTesting f544ec76c810ba735b5d1492c493e84cfe81a719 == 43dc4f85a9519e9c93e65cba75271725a8f17e51", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-141.html", "line": 9729, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/CSS2/fonts/font-family-invalid-characters-003.xht a0e064dd962fcefc258d77ac4e104111af399d43\n/css/CSS2/fonts/font-family-invalid-characters-002-ref.html 0312ee62c35e20c1dbabe5967b41e73fad2c51bf\nTesting a0e064dd962fcefc258d77ac4e104111af399d43 == 0312ee62c35e20c1dbabe5967b41e73fad2c51bf", "stack": null, "subtest": null, "test": "/css/CSS2/fonts/font-family-invalid-characters-003.xht", "line": 13904, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/CSS2/fonts/font-family-rule-001.xht b56b2d42a123bfbc2b004916b32325a93db34d89\n/css/CSS2/fonts/font-family-rule-001-ref.html 2c8e2dd919cce243d53e3362ea886c8341a087a1\nTesting b56b2d42a123bfbc2b004916b32325a93db34d89 == 2c8e2dd919cce243d53e3362ea886c8341a087a1", "stack": null, "subtest": null, "test": "/css/CSS2/fonts/font-family-rule-001.xht", "line": 13928, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/CSS2/fonts/fonts-012.xht 4ec417481cf44f06f2e99d3a54e862601cfc9414\n/css/CSS2/fonts/fonts-012-ref.xht a3198c142f82386daa3a15a566dd0e7357c46716\nTesting 4ec417481cf44f06f2e99d3a54e862601cfc9414 == a3198c142f82386daa3a15a566dd0e7357c46716", "stack": null, "subtest": null, "test": "/css/CSS2/fonts/fonts-012.xht", "line": 14294, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-016.html 0c19462ca8706e5bf9a66ff1c43fcf6647481820\n/css/css-text/i18n/reference/css3-text-line-break-opclns-016-ref.html 5e041cea003b2e64c441176603dd7f53a3db44c3\nTesting 0c19462ca8706e5bf9a66ff1c43fcf6647481820 == 5e041cea003b2e64c441176603dd7f53a3db44c3", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-016.html", "line": 20382, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-023.html 11e8e7b33a48a75e8b900f448101fd53716ca28e\n/css/css-text/i18n/reference/css3-text-line-break-opclns-023-ref.html 25173f7bf2f0f2c1b62e5a05b2f2d1831fbb6969\nTesting 11e8e7b33a48a75e8b900f448101fd53716ca28e == 25173f7bf2f0f2c1b62e5a05b2f2d1831fbb6969", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-023.html", "line": 20410, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-027.html 2c429196f2945693bb76bc27cf3345beb3c10200\n/css/css-text/i18n/reference/css3-text-line-break-opclns-027-ref.html e53a44579f21d63231e9ed1f45f38bb827626420\nTesting 2c429196f2945693bb76bc27cf3345beb3c10200 == e53a44579f21d63231e9ed1f45f38bb827626420", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-027.html", "line": 20432, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-030.html a05ead8455d62bec7124c42a834b599f091f6585\n/css/css-text/i18n/reference/css3-text-line-break-opclns-030-ref.html a3fec18d5a5436c652f6ef3e13cdbe9059713cb8\nTesting a05ead8455d62bec7124c42a834b599f091f6585 == a3fec18d5a5436c652f6ef3e13cdbe9059713cb8", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-030.html", "line": 20454, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-034.html adce5d7e9bc4fe16e7e4837cc044b73311325d4c\n/css/css-text/i18n/reference/css3-text-line-break-opclns-034-ref.html eea0168da64ff1ef6db34d82241c22b9f83ee799\nTesting adce5d7e9bc4fe16e7e4837cc044b73311325d4c == eea0168da64ff1ef6db34d82241c22b9f83ee799", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-034.html", "line": 20476, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-006.html 50cb514fb62b1ecc5dd0389ccd9232037108ba31\n/css/css-text/i18n/reference/css3-text-line-break-opclns-006-ref.html 7a95d4de66ac732dc2461cc9ae6ef23ab1f4d937\nTesting 50cb514fb62b1ecc5dd0389ccd9232037108ba31 == 7a95d4de66ac732dc2461cc9ae6ef23ab1f4d937", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-006.html", "line": 20689, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-015.html d3813e6948598b0f99ef5af4ef758ae86b66643d\n/css/css-text/i18n/reference/css3-text-line-break-opclns-015-ref.html 1e120f284fee1b184d4cfc629d67df7512d5303e\nTesting d3813e6948598b0f99ef5af4ef758ae86b66643d == 1e120f284fee1b184d4cfc629d67df7512d5303e", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-015.html", "line": 20734, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-122.html 240ea0e620394c0e3a65f19a0d5e3999907994ee\n/css/css-text/i18n/reference/css3-text-line-break-opclns-122-ref.html 77638e4fe05ad7a12fc61121d80e8e7ef0d06e54\nTesting 240ea0e620394c0e3a65f19a0d5e3999907994ee == 77638e4fe05ad7a12fc61121d80e8e7ef0d06e54", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-122.html", "line": 20743, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-019.html 32372ccedcb68a1d130b6ca53dc725067fa65837\n/css/css-text/i18n/reference/css3-text-line-break-opclns-019-ref.html 300492703c23fb73cac9dd948f5e3b9df51aa031\nTesting 32372ccedcb68a1d130b6ca53dc725067fa65837 == 300492703c23fb73cac9dd948f5e3b9df51aa031", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-019.html", "line": 20758, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-020.html 9cad1d9d5cc54e452f92997c2698137b3e1d2825\n/css/css-text/i18n/reference/css3-text-line-break-opclns-020-ref.html 672b24f848bbe69c0d4fd089039e49a4a8cabd86\nTesting 9cad1d9d5cc54e452f92997c2698137b3e1d2825 == 672b24f848bbe69c0d4fd089039e49a4a8cabd86", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-020.html", "line": 20777, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-131.html 4607d003c5075d1f5197b864a40baddb06015af6\n/css/css-text/i18n/reference/css3-text-line-break-opclns-131-ref.html ebd33ab5894cc40de0f4dc6269ec0e825c5b4c88\nTesting 4607d003c5075d1f5197b864a40baddb06015af6 == ebd33ab5894cc40de0f4dc6269ec0e825c5b4c88", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-131.html", "line": 20788, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-135.html bcfd25f47f1445b1f1b9e8a77ec2c18c16d2fe46\n/css/css-text/i18n/reference/css3-text-line-break-opclns-135-ref.html fb3a89bbf8104ffb9d72b4dc90c52bf210d214bc\nTesting bcfd25f47f1445b1f1b9e8a77ec2c18c16d2fe46 == fb3a89bbf8104ffb9d72b4dc90c52bf210d214bc", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-135.html", "line": 20812, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-028.html 5b91fe218f1745579a76b7942aee78da23378027\n/css/css-text/i18n/reference/css3-text-line-break-opclns-028-ref.html 6f78263aa11d2ec7057d4e8e5393c8d856b7981d\nTesting 5b91fe218f1745579a76b7942aee78da23378027 == 6f78263aa11d2ec7057d4e8e5393c8d856b7981d", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-028.html", "line": 20825, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-139.html b5ad1332811da67be05050b57b660bd57a0481c0\n/css/css-text/i18n/reference/css3-text-line-break-opclns-139-ref.html fd2c862c038e853bdc4d657ebb4c305d62264092\nTesting b5ad1332811da67be05050b57b660bd57a0481c0 == fd2c862c038e853bdc4d657ebb4c305d62264092", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-139.html", "line": 20837, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-033.html 933d6d92835e79241378a9b2c38a1ced496cb8e1\n/css/css-text/i18n/reference/css3-text-line-break-opclns-033-ref.html b12f062f799f683d672c6f776f4df4230cfddeb9\nTesting 933d6d92835e79241378a9b2c38a1ced496cb8e1 == b12f062f799f683d672c6f776f4df4230cfddeb9", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-033.html", "line": 20849, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-140.html 1e4fa560f53352e48da0fc717473c43d2b3a2ccd\n/css/css-text/i18n/reference/css3-text-line-break-opclns-140-ref.html 16ce4bcce8320030adff7d8102414828c6a2c3aa\nTesting 1e4fa560f53352e48da0fc717473c43d2b3a2ccd == 16ce4bcce8320030adff7d8102414828c6a2c3aa", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-140.html", "line": 20861, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-037.html fdf3720ffe44ea00737bdfe154ad9fa454c159bf\n/css/css-text/i18n/reference/css3-text-line-break-opclns-037-ref.html ab811c3e542ccac3ccdd55f029b2cd3353bb62f0\nTesting fdf3720ffe44ea00737bdfe154ad9fa454c159bf == ab811c3e542ccac3ccdd55f029b2cd3353bb62f0", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-037.html", "line": 20876, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-125.html 2f14d1ed02056a3efba19b3205bc832f1aa1f30f\n/css/css-text/i18n/reference/css3-text-line-break-opclns-125-ref.html 6d735a248ca5e0d2c0d8540f528ffed4e7f2785b\nTesting 2f14d1ed02056a3efba19b3205bc832f1aa1f30f == 6d735a248ca5e0d2c0d8540f528ffed4e7f2785b", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-125.html", "line": 21140, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-132.html 3ec6bd9b3a131b1c2980f699a8967839fb8600fb\n/css/css-text/i18n/reference/css3-text-line-break-opclns-132-ref.html 4cc85bd9be95d896c1128e4e6ddc5bbb776b00cd\nTesting 3ec6bd9b3a131b1c2980f699a8967839fb8600fb == 4cc85bd9be95d896c1128e4e6ddc5bbb776b00cd", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-132.html", "line": 21193, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/css/css-text/i18n/css3-text-line-break-opclns-136.html 0d8ba3d9b1174fb0078dbb969c939ad396e30967\n/css/css-text/i18n/reference/css3-text-line-break-opclns-136-ref.html dc174dc1ce690ac67d6fc3be14b4da1f197b0465\nTesting 0d8ba3d9b1174fb0078dbb969c939ad396e30967 == dc174dc1ce690ac67d6fc3be14b4da1f197b0465", "stack": null, "subtest": null, "test": "/css/css-text/i18n/css3-text-line-break-opclns-136.html", "line": 21220, "action": "test_result", "expected": "PASS"}
```

</details>

bors-servo commented:
:hourglass: Testing commit 0f7b6336d6bee534305461361ab3436d0c2d319e with merge f98c66406822bac83f562c19797ca99a38681144...

jonleighton commented:
> linux-rel-wpt
>
>     {"status": "FAIL", "group": "default", "message": "/_mozilla/css/font_fallback_01.html 933650d419a5c77b9aaf59d983187ab712b75312\n/_mozilla/css/font_fallback_failed_ref.html 933650d419a5c77b9aaf59d983187ab712b75312\nTesting 933650d419a5c77b9aaf59d983187ab712b75312 != 933650d419a5c77b9aaf59d983187ab712b75312", "stack": null, "subtest": null, "test": "/_mozilla/css/font_fallback_01.html", "line": 42512, "action": "test_result", "expected": "PASS"}
>     {"status": "FAIL", "group": "default", "message": "/_mozilla/css/font_fallback_02.html 933650d419a5c77b9aaf59d983187ab712b75312\n/_mozilla/css/font_fallback_failed_ref.html 933650d419a5c77b9aaf59d983187ab712b75312\nTesting 933650d419a5c77b9aaf59d983187ab712b75312 != 933650d419a5c77b9aaf59d983187ab712b75312", "stack": null, "subtest": null, "test": "/_mozilla/css/font_fallback_02.html", "line": 49681, "action": "test_result", "expected": "PASS"}

These two failures are because we're running these tests (which are added by this PR) on a server machine which hasn't got default fonts to render these characters.

The test inherently relies on various system fonts being present, which in hindsight is maybe not such a great idea? Solutions I can think of:

1. Configure Linux builders to install certain fonts (downside: somebody running the tests on their own setup may get failures due to missing fonts)
2. Remove these tests (downside: this functionality could regress, though I might be able to cook something up in `components/gfx/tests`)

Any opinions?

I will look into the rest of the failures...

jonleighton commented:
> (Also a number of other unexpected results that got filtered out as intermittent but might be relevant, since several of them are in text layout tests.)

The `css-text` ones also show up on current master, so I don't think they're affected by this change. (Although they seem to reliably pass for me, so I don't know why they're marked intermittent...)

jonleighton commented:
OK:

* I've checked through the Mac failures and generally just updated the metadata accordingly. Generally the pass/fail state has changed because we are now able to render a character which we previously wouldn't. I've added an explanation of the metadata changes to [the commit](https://github.com/servo/servo/pull/20506/commits/4d0cf0d2c57ca379c792c23fdf45bdbdfeaecf74)
* I've added a unit test for fallback behaviour
* I've left the fallback reftests in there, but marked 01 and 02 failing on Linux. These can start to pass if we install the appropriate fonts on the builders, which I'd like to look at as a follow up. (I debated this, but there are a number of existing tests which will never work properly unless the builder has certain fonts, so it seems reasonable to do so...)
* I changed the code so that [the first fallback option we consider is the generic `serif` font](https://github.com/servo/servo/pull/20506/files#diff-3d35bda7ed0efd25268c5ae0851f0e33R416). This fixes several tests that rely on `font-family: AFontNameThatDoesntExist` rendering the default font (i.e. whatever `serif` resolves to). Without this tests would fail because the first fallback font on mac was a sans-serif (Lucida Grande).

jonleighton commented:
@mbrubeck could you give this another r+ if you're happy with the above?

mbrubeck commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit d9618ec has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck d9618ec17f025872f4725c6e5ed0b1c613b31367 -->

bors-servo commented:
:hourglass: Testing commit d9618ec17f025872f4725c6e5ed0b1c613b31367 with merge a851f83ea47b14efb0b6668d768623c385af743e...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8329)

jonleighton commented:
:cry: 

These two failed:

* http://build.servo.org/builders/linux-rel-nogate/builds/2193
* http://build.servo.org/builders/mac-rel-wpt1/builds/8329

The first one is unrelated to this change, but given the name (`linux-rel-nogate`) I guess it doesn't need to be passing?

The second is something to do with intermittents. I don't know much about how intermittents work so I'll need to look into it.

bors-servo commented:
:hourglass: Testing commit d9618ec17f025872f4725c6e5ed0b1c613b31367 with merge d12a023329d2cc5ba019aaeecd1800cd0672ced5...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8361)

jonleighton commented:
Ok, this is the one I hope! :crossed_fingers:  :crossed_fingers:  :crossed_fingers:

emilio commented:
@bors-servo try

 * You should get some try access btw ;)

bors-servo commented:
:hourglass: Trying commit d9c24b0fe34639ea44df67770ac2e750134e85fc with merge 5efcdb8f3c30dceae04a7bba5ccde0ba6aae9c59...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11740)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20420) made this pull request unmergeable. Please resolve the merge conflicts.

jonleighton commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 63041702f73351b4681b9d7fd2d865ad48c4a320 with merge 48217d62a8fddbefcb4d6eda3f609574bdc6f6d7...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8684)

CYBAI commented:
```rust
error[E0277]: the trait bound `font_template::FontTemplateDescriptor: std::cmp::Eq` is not satisfied
   --> components/gfx/font.rs:123:5
    |
123 |     pub template_descriptor: FontTemplateDescriptor,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::cmp::Eq` is not implemented for `font_template::FontTemplateDescriptor`
    |
    = note: required by `std::cmp::AssertParamIsEq`

error[E0277]: the trait bound `font_template::FontTemplateDescriptor: std::hash::Hash` is not satisfied
   --> components/gfx/font.rs:123:5
    |
123 |     pub template_descriptor: FontTemplateDescriptor,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::hash::Hash` is not implemented for `font_template::FontTemplateDescriptor`
    |
    = note: required by `std::hash::Hash::hash`

error[E0277]: the trait bound `font_template::FontTemplateDescriptor: std::cmp::Eq` is not satisfied
   --> components/gfx/font_context.rs:206:5
    |
206 |     template_descriptor: FontTemplateDescriptor,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::cmp::Eq` is not implemented for `font_template::FontTemplateDescriptor`
    |
    = note: required by `std::cmp::AssertParamIsEq`

error[E0277]: the trait bound `font_template::FontTemplateDescriptor: std::hash::Hash` is not satisfied
   --> components/gfx/font_context.rs:206:5
    |
206 |     template_descriptor: FontTemplateDescriptor,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::hash::Hash` is not implemented for `font_template::FontTemplateDescriptor`
    |
    = note: required by `std::hash::Hash::hash`

error: aborting due to 4 previous errors
```

CYBAI commented:
@jonleighton Just help to summarize the error :P

jonleighton commented:
@emilio I'm not sure what to do to fix this. The branch is now failing to build because [`FontDescriptor` derives `Eq` and `Hash`](https://github.com/servo/servo/pull/20506/files#diff-3d35bda7ed0efd25268c5ae0851f0e33R121), which means that `FontTemplateDescriptor` must also implement `Eq` and `Hash`. This was the case when I wrote this code, but it looks like some of your more recent work has changed that (e.g. 48de556f8caebe1c09b9d2542b074720ff76e8bc). In general it is now not trivial to just derive them again, because some of the nested structs (such as `FontWeight`) are wrapping around floats now. Do you have any thoughts about the best way forward? I noticed that you did at one point [implement `Hash` for `FontWeight`](https://github.com/servo/servo/commit/48de556f8caebe1c09b9d2542b074720ff76e8bc#diff-6cb05dd2a8bb492bb45afa6895d6908cR125) although it doesn't look like that code is there any more. Maybe we could bring it back?

BTW, the reason my code needs `Hash` is because `FontDescriptor` and `FontTemplateDescriptor` are used in hash keys [here](https://github.com/servo/servo/pull/20506/files#diff-55e83391bc898f82b9b8d0dfe4c16b1fR52).

emilio commented on components/gfx/font.rs:
```diff
@@ -352,35 +358,78 @@ impl FontGroup {
+        S: FontSource,
+        P: FnMut(&FontRef) -> bool,
+    {
+        vec![FontFamilyDescriptor::default()].into_iter()
```
nit: iter::once?

jonleighton replied:
Thanks, didn't know about that one

emilio commented on components/gfx/font_context.rs:
```diff
@@ -160,76 +103,88 @@ impl<S: FontSource> FontContext<S> {
+        font_descriptor: &FontDescriptor,
+        family_descriptor: &FontFamilyDescriptor,
+    ) -> Option<FontRef>
+    {
```
nit: brace to the previous line.

jonleighton replied:
Thanks, I thought this was the style [recommended here](https://github.com/rust-lang-nursery/fmt-rfcs/blob/master/guide/items.md) but I've realised they only do a line break when there is a `where` clause.

emilio commented on components/gfx/font_context.rs:
```diff
@@ -160,76 +103,88 @@ impl<S: FontSource> FontContext<S> {
+        template_descriptor: &FontTemplateDescriptor,
+        family_descriptor: &FontFamilyDescriptor
+    ) -> Option<FontTemplateInfo>
+    {
```
nit: brace.

emilio commented on components/gfx/font_context.rs:
```diff
@@ -160,76 +103,88 @@ impl<S: FontSource> FontContext<S> {
+        info: FontTemplateInfo,
+        descriptor: FontDescriptor
+    ) -> Result<Font, ()>
+    {
```
brace

emilio commented on components/gfx/font_cache_thread.rs:
```diff
@@ -320,44 +316,54 @@ impl FontCache {
+        template_descriptor: &FontTemplateDescriptor,
+        family_name: &FontFamilyName,
+    ) -> Option<Arc<FontTemplateData>>
+    {
```
nit: brace

emilio reviewed (commented):
So, implementing `Hash` on `FontWeight` should be fine, since it has a range guarantee. But we'd need to hash font-style and font-stretch too, which are also floats now.

Instead of doing a bunch of wrapper types (since font-stretch is a `NonNegativePercentage`), I decided to put the hashing in `compute_font_hash`.

Looks like that code is missing font-style and I noticed.

In any case, given `compute_font_hash` and `FontTemplateDescriptor`'s hashes are supposed to be in sync, we may as well make it so moving the hashing to a common place.

What about moving most of the bits of `compute_font_hash` to `style/values/computed/font.rs`, make that a function that takes a `Hasher`, `FontWeight`, `FontStretch` and `FontStyle`, and `impl Hash for FontTemplateDescriptor` using that function as well?

That would solve `Hash`. Now for `Eq`... The only reason floats aren't `Eq` is because of `NaN`. We shouldn't be able to get `NaN` in any of the fields of `FontTemplateDescriptor`, so manually implementing it should just work?

Hope it helps. Thanks for working on this!

jonleighton commented:
> Hope it helps. Thanks for working on this!

It took me a few reads, but I think I do understand what you're proposing. I'll have a go at implementing when I have a spare moment :+1:

jonleighton commented:
@emilio do you reckon https://github.com/servo/servo/pull/20506/commits/c8ccf8804b29ab15f91a0c7885791783eb656a64 is looking on the right track? I still have some build errors to fix there, but would be good to know if you think that direction looks OK.

jonleighton commented:
@emilio ok, it's building now ^^ If you're happy with the above then I'll squash it into the original commit and then see if I can finally get WPT to pass!

emilio commented on components/style/values/generics/font.rs:
```diff
@@ -228,7 +228,7 @@ impl Default for KeywordSize {
 #[allow(missing_docs)]
 #[cfg_attr(feature = "servo", derive(Deserialize, Serialize))]
-#[derive(Animate, Clone, ComputeSquaredDistance, Copy, Debug, MallocSizeOf,
+#[derive(Animate, Clone, ComputeSquaredDistance, Copy, Debug, MallocSizeOf, Hash,
```
nit: tidy won't be happy about this.

jonleighton replied:
Tidy doesn't care (I have the pre-commit hook setup), I think its line limit is longer than this line. Happy to wrap it to the next line though.

emilio replied:
Oh, I meant because of the derive alphabetic sorting... Looks like a bug in tidy? Oh well :)

jonleighton replied:
Aha! Well, fixed now anyway.

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -2305,13 +2305,10 @@ pub mod style_structs {
+                    self.font_weight.hash(&mut hasher);
+                    self.font_stretch.hash(&mut hasher);
+                    self.font_style.hash(&mut hasher);
                     self.font_family.hash(&mut hasher);
```
I suspect we shouldn't be serializing font_family... But maybe worth filing a follow-up?

jonleighton replied:
:+1: https://github.com/servo/servo/issues/20787

emilio reviewed (commented):
The style changes look fine to me, thank you!

jonleighton commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit be48d84d3dbe8291e5f752d7ea13ba3d3705ca45 with merge abfacfc1bfed0dfb73de82ecddbbc1fcad3bf8d7...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/11503)

jdm commented:
@bors-servo try

Eijebong commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit be48d84d3dbe8291e5f752d7ea13ba3d3705ca45 with merge 5d363e102555cb9f0872a9f1304e71391f4ec9e0...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8579)

jonleighton commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 3025a269ec47240226cdd83f7138c4e61ca379f0 with merge 30212d9db87bdd1502e7a1ef4d9c8ac0518b921f...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8764)

jonleighton commented:
@bors-servo retry

Reason: seems like a random crash:

```
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/history/the-location-interface/location-protocol-setter.html", "line": 96789, "action": "test_result", "expected": "OK"}
```

bors-servo commented:
:hourglass: Trying commit 3025a269ec47240226cdd83f7138c4e61ca379f0 with merge 1b56bf5991c130162f724de0c935583bb9f821c8...

jdm commented:
Rather than immediately retrying, it's best to rely on [the grid](http://build.servo.org/grid) to see what other test results are reported. In particular, retrying doesn't cancel any ongoing test runs so it also delays the rest of the queue.

jonleighton commented:
Ok, noted, I didn't realise that the previous build hadn't completely finished.

jonleighton commented:
Ok, noted, I didn't realise that the previous build hadn't completely finished.

jonleighton commented:
Ok, noted, I didn't realise that the previous build hadn't completely finished.

jonleighton commented:
Ok, the retry worked! (Maybe it didn't report here as I accidentally closed and reopened the PR...)

Can I get a merge please? :pray: :smile:

emilio commented:
Homu is stuck on this PR, just trying to un-stick it.

@bors-servo r-

bors-servo commented:
:hourglass: Testing commit 3025a269ec47240226cdd83f7138c4e61ca379f0 with merge 824f40319b4915c81543fc7ccdc017f9ad582255...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4341)

jonleighton commented:
o_O we managed to crash `rustc`:

```
rustc(40975,0x700000e13000) malloc: *** error for object 0x1: pointer being freed was not allocated
```

bors-servo commented:
:hourglass: Testing commit 3025a269ec47240226cdd83f7138c4e61ca379f0 with merge 77dcc678febed8a7859901d6b925d368c6fe8953...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/11559), [arm32](http://build.servo.org/builders/arm32/builds/11785), [arm64](http://build.servo.org/builders/arm64/builds/11833), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11873), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8773), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8776), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11931), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4325), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4349), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8597), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8647), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4418), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4347), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7264)
Approved by: emilio,mbrubeck
Pushing 77dcc678febed8a7859901d6b925d368c6fe8953 to master...
